File: sync_db.py
#!/usr/bin/env python3
"""Sync likes/dislikes counts from Redis sets into Postgres videos table.

Usage:
  python sync_db.py            # uses REDIS_URL and DATABASE_URL / PG* env vars
  python sync_db.py --dry-run  # do not write changes, just print
"""
import os
import re
import argparse
import redis
import psycopg2
import psycopg2.extras

KEY_RE = re.compile(r'^video:(?P<id>\d+):(?P<kind>likes|dislikes)$')


def get_pg_connection():
    try:
        dsn = os.environ.get('DATABASE_URL')
        if dsn:
            conn = psycopg2.connect(dsn, cursor_factory=psycopg2.extras.RealDictCursor)
        else:
            host = os.environ.get('PGHOST', 'localhost')
            port = int(os.environ.get('PGPORT', '5432'))
            user = os.environ.get('PGUSER', 'postgres')
            password = os.environ.get('PGPASSWORD', '')
            dbname = os.environ.get('PGDATABASE', 'smartvideos')
            conn = psycopg2.connect(host=host, port=port, user=user, password=password, dbname=dbname,
                                    cursor_factory=psycopg2.extras.RealDictCursor)
        return conn
    except Exception as e:
        raise RuntimeError(f'Postgres connection error: {e}')


def main(redis_url, dry_run=False, pattern='video:*:likes'):
    r = redis.from_url(redis_url)

    # We'll scan for both likes and dislikes keys
    cursor = 0
    seen = 0
    try:
        conn = get_pg_connection()
    except Exception as e:
        print('ERROR: cannot connect to Postgres:', e)
        return 1

    try:
        print('Scanning Redis for keys matching video:*:likes and video:*:dislikes...')
        for key in r.scan_iter(match='video:*:likes'):
            key = key.decode() if isinstance(key, bytes) else key
            m = KEY_RE.match(key)
            if not m:
                print('Skipping non-matching key:', key)
                continue
            vid = int(m.group('id'))
            kind = m.group('kind')
            count = int(r.scard(key) or 0)
            print(f'Key={key} -> video_id={vid} {kind} count={count}')
            if not dry_run:
                try:
                    cur = conn.cursor()
                    cur.execute('UPDATE videos SET likes_count = %s WHERE id = %s', (count, vid))
                    conn.commit()
                    cur.close()
                except Exception as e:
                    print('DB update error for video', vid, 'likes:', e)
            seen += 1

        for key in r.scan_iter(match='video:*:dislikes'):
            key = key.decode() if isinstance(key, bytes) else key
            m = KEY_RE.match(key)
            if not m:
                print('Skipping non-matching key:', key)
                continue
            vid = int(m.group('id'))
            kind = m.group('kind')
            count = int(r.scard(key) or 0)
            print(f'Key={key} -> video_id={vid} {kind} count={count}')
            if not dry_run:
                try:
                    cur = conn.cursor()
                    cur.execute('UPDATE videos SET dislikes_count = %s WHERE id = %s', (count, vid))
                    conn.commit()
                    cur.close()
                except Exception as e:
                    print('DB update error for video', vid, 'dislikes:', e)
            seen += 1

        print('Done. Processed', seen, 'keys.')
    finally:
        try:
            conn.close()
        except Exception:
            pass

    return 0


if __name__ == '__main__':
    p = argparse.ArgumentParser(description='Sync Redis video likes/dislikes counts into Postgres')
    p.add_argument('--redis', dest='redis_url', default=os.environ.get('REDIS_URL', 'redis://localhost:6379/0'), help='Redis URL')
    p.add_argument('--dry-run', dest='dry_run', action='store_true', help='Do not write to the database')
    args = p.parse_args()
    raise SystemExit(main(args.redis_url, dry_run=args.dry_run))

File: sync_db.go
package main

import (
	"context"
	"database/sql"
	"flag"
	"fmt"
	"log"
	"os"
	"regexp"
	"strconv"

	_ "github.com/jackc/pgx/v5/stdlib"
	"github.com/redis/go-redis/v9"
)

var (
	syncCtx = context.Background()
	keyRe   = regexp.MustCompile(`^video:(?P<id>\d+):(?P<kind>likes|dislikes)$`)
)

func getEnvString(key, fallback string) string {
	if val := os.Getenv(key); val != "" {
		return val
	}
	return fallback
}

func getPGConn() *sql.DB {
	dsn := os.Getenv("DATABASE_URL")
	if dsn == "" {
		host := getEnvString("PGHOST", "localhost")
		port := getEnvString("PGPORT", "5432")
		user := getEnvString("PGUSER", "postgres")
		password := os.Getenv("PGPASSWORD")
		dbname := getEnvString("PGDATABASE", "smartvideos")
		dsn = fmt.Sprintf("postgres://%s:%s@%s:%s/%s?sslmode=disable", user, password, host, port, dbname)
	}

	database, err := sql.Open("pgx", dsn)
	if err != nil {
		log.Fatalf("[FATAL] Erro ao abrir conexão com o Postgres: %v", err)
	}
	return database
}

func processKeys(rdb *redis.Client, db *sql.DB, pattern string, isDryRun bool, column string) int {
	var cursor uint64
	processed := 0

	for {
		keys, nextCursor, err := rdb.Scan(syncCtx, cursor, pattern, 0).Result()
		if err != nil {
			log.Printf("[ERRO] Erro ao escanear o Redis com o padrão %s: %v", pattern, err)
			return processed
		}

		for _, key := range keys {
			matches := keyRe.FindStringSubmatch(key)
			if len(matches) < 3 {
				log.Printf("[INFO] Pulando chave fora do padrão: %s", key)
				continue
			}

			videoIDStr := matches[1]
			videoID, err := strconv.Atoi(videoIDStr)
			if err != nil {
				continue
			}

			count, err := rdb.SCard(syncCtx, key).Result()
			if err != nil {
				log.Printf("[ERRO] Erro ao ler contagem (SCard) da chave %s: %v", key, err)
				continue
			}

			fmt.Printf("Chave=%s -> video_id=%d contagem de %s=%d\n", key, videoID, column, count)

			if !isDryRun {
				query := fmt.Sprintf("UPDATE videos SET %s = $1 WHERE id = $2", column)
				_, err := db.Exec(query, count, videoID)
				if err != nil {
					log.Printf("[ERRO] Erro ao atualizar o banco para o vídeo %d (%s): %v", videoID, column, err)
				}
			}
			processed++
		}

		cursor = nextCursor
		if cursor == 0 {
			break
		}
	}

	return processed
}

func main() {
	redisURLFlag := flag.String("redis", getEnvString("REDIS_URL", "redis://localhost:6379/0"), "URL do Redis")
	dryRunFlag := flag.Bool("dry-run", false, "Não grava as alterações no banco, apenas printa")
	flag.Parse()

	// 1. Inicializa Conexão com o Redis
	opt, err := redis.ParseURL(*redisURLFlag)
	if err != nil {
		log.Fatalf("[FATAL] Erro ao processar URL do Redis: %v", err)
	}
	rdb := redis.NewClient(opt)
	defer rdb.Close()

	if err := rdb.Ping(syncCtx).Err(); err != nil {
		log.Fatalf("[FATAL] Não foi possível conectar ao Redis: %v", err)
	}

	// 2. Inicializa Conexão com o Postgres
	db := getPGConn()
	defer db.Close()

	log.Println("⚡ Escaneando Redis para sincronizar likes e dislikes com o Postgres...")

	totalSeen := 0
	totalSeen += processKeys(rdb, db, "video:*:likes", *dryRunFlag, "likes_count")
	totalSeen += processKeys(rdb, db, "video:*:dislikes", *dryRunFlag, "dislikes_count")

	fmt.Printf("✅ Sincronização concluída! Processadas %d chaves com sucesso.\n", totalSeen)
	os.Exit(0)
}
